safeweb/http.go

```diff
@@ -206,24 +221,22 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	case bp == "" && ap == "": // neither match
 		http.NotFound(w, r)
 	case bp != "" && ap != "":
-		// Both muxes match the path. This can be because:
-		//  * one of them registers a wildcard "/" handler
-		//  * there are overlapping specific handlers
-		//
-		// If it's the former, route to the more-specific handler. If it's the
-		// latter - that's a bug so return an error to avoid mis-routing the
-		// request.
+		// Both muxes match the path. Route to the more-specific handler (as
+		// determined by the number of components in the path). If it somehow
+		// happens that both patterns are equally specific, something strange
+		// has happened; say so.
 		//
-		// TODO(awly): match the longest path instead of only special-casing
-		// "/".
-		switch {
-		case bp == "/":
+		// NOTE: getMoreSpecificPattern does not know about what the serve*
+		// handlers will do — including, possibly, redirecting to more specific
+		// patterns. If you have a less-specific pattern that redirects to
+		// something more specific, this logic will not do what you wanted.
+		specific := getMoreSpecificPattern(ap, bp)
```

Comment:
optional: use `switch getMoreSpecificPattern(ap, bp)`

Reply:
Seeing the usage, maybe instead of reporting "which one", we really want a little enumeration:

```go
type handlerType int

const (
   unknownHandler handlerType = iota
   apiHandler
   browserHandler
)

func checkHandlerType(apiPattern, browserPattern string) handlerType { … }

switch checkHandlerType(ap, bp) {
case apiHandler:
   s.serveAPI(w, r)
case browserHandler:
   s.serveBrowser(w, r)
default
   // report error, etc.
}
```

Not a blocker, but I feel like that might express intent more clearly. That said, what you have here is OK too.